Provider/src/main/java/com/spotify/confidence/Confidence.kt

```diff
@@ -0,0 +1,185 @@
+    private val eventSenderEngine: EventSenderEngine,
+    private val diskStorage: DiskStorage,
+    private val flagResolver: FlagResolver,
+    private val flagApplierClient: FlagApplierClient,
+    private val parent: ConfidenceContextProvider? = null,
+    private val region: ConfidenceRegion = ConfidenceRegion.GLOBAL
+) : Confidence by ConfidenceImpl(
+    clientSecret,
+    dispatcher,
+    eventSenderEngine,
+    diskStorage,
+    flagResolver,
+    flagApplierClient,
+    parent,
+    region
+)
+
+class ConfidenceImpl internal constructor(
+    private val clientSecret: String,
+    private val dispatcher: CoroutineDispatcher,
+    private val eventSenderEngine: EventSenderEngine,
+    private val diskStorage: DiskStorage,
+    private val flagResolver: FlagResolver,
+    private val flagApplierClient: FlagApplierClient,
+    private val parent: ConfidenceContextProvider? = null,
+    private val region: ConfidenceRegion = ConfidenceRegion.GLOBAL
+) : Confidence {
+    private val removedKeys = mutableListOf<String>()
+    private var contextMap: MutableMap<String, ConfidenceValue> = mutableMapOf()
+
+    override fun putContext(key: String, value: ConfidenceValue) {
+        contextMap[key] = value
+    }
+
+    override fun putContext(context: Map<String, ConfidenceValue>) {
+        contextMap += context
+    }
+
+    override fun setContext(context: Map<String, ConfidenceValue>) {
+        contextMap = context.toMutableMap()
+    }
+
+    override fun removeContext(key: String) {
+        removedKeys.add(key)
+        contextMap.remove(key)
+    }
+
+    override fun getContext(): Map<String, ConfidenceValue> =
+        this.parent?.let {
+            it.getContext().filterKeys { key -> !removedKeys.contains(key) } + contextMap
+        } ?: contextMap
+
+    override fun withContext(context: Map<String, ConfidenceValue>): Confidence = ChildConfidence(
+        clientSecret,
+        dispatcher,
+        eventSenderEngine,
+        diskStorage,
+        flagResolver,
+        flagApplierClient,
+        this,
+        region
+    ).also {
+        it.putContext(context)
+    }
+
+    override fun send(
+        definition: String,
+        payload: ConfidenceFieldsType
+    ) {
+        eventSenderEngine.emit(definition, payload, getContext())
+    }
+}
+
+internal fun Map<String, ConfidenceValue>.openFeatureFlatten(): Map<String, ConfidenceValue> {
+    val context = this.toMutableMap()
+    val openFeatureContext = context[OPEN_FEATURE_CONTEXT_KEY]?.let { it as ConfidenceValue.Struct }
+    openFeatureContext?.let {
+        context += it.map
+    }
+    context.remove(OPEN_FEATURE_CONTEXT_KEY)
+    return context
+}
+
+internal const val OPEN_FEATURE_CONTEXT_KEY = "open_feature"
+
+object ConfidenceFactory {
+    fun create(
+        context: Context,
+        clientSecret: String,
+        region: ConfidenceRegion = ConfidenceRegion.GLOBAL,
+        dispatcher: CoroutineDispatcher = Dispatchers.IO
+    ): RootConfidence {
```

Comment:
What do you think about returning `Confidence` here and have `Confidence` also include the `shutdown()`.
However we keep the two different types (`RootConfidence` and `ChildConfidence`) but running `shutdown()` on a child is a no-op (only log).

Reply:
then we can remove root and child confidence and have shutdown in confidence always. right now it could be no-op but in future it could de-register child and all children are de-registered then the engine will stop.